include/unifex/exception.hpp

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <exception>
+#include <utility>
+#include <type_traits>
+
+#include <unifex/std_concepts.hpp>
+
+#include <unifex/detail/prologue.hpp>
+
+// Microsoft's STL has a fast implementation of std::make_exception_ptr.
+// So does libc++ when targetting Microsoft's ABI. On other stdlibs,
+// move as much exception machinery into a cpp file as possible to avoid
+// template code bloat.
+#if !defined(_LIBCPP_ABI_MICROSOFT) && !defined(_MSVC_STL_VERSION)
+#define UNIFEX_HAS_FAST_MAKE_EXCEPTION_PTR() 0
+#else
+#define UNIFEX_HAS_FAST_MAKE_EXCEPTION_PTR() 1
+#endif
+
+namespace unifex {
+namespace _except_ptr {
+
+#if !UNIFEX_HAS_FAST_MAKE_EXCEPTION_PTR()
+// If std::make_exeption_ptr() is slow, then move it into a cpp
+// file to generate less code elsewhere.
+struct _ref {
+  template (typename Obj)
+    (requires (!same_as<remove_cvref_t<Obj>, _ref>))
+  _ref(Obj&& obj) noexcept
+    : p_((void*) std::addressof(obj))
+    , throw_([](void* p) {
+        throw (Obj&&) *(std::add_pointer_t<Obj>) p;
```

Comment:
This probably needs some `#ifdef` around it to support `-fno-exceptions`..

Reply:
There's a bunch of places in the code where we need to do something special for `-fno-exceptions`. It's unclear to me what that thing should be here, and answering that question is not the point of this PR. Unifex doesn't support `-fno-exceptions` yet. TBD.

Reply:
I thought @ispeters had done a pass through the code to make `-fno-exceptions` work, at least for a subset of the code...

Reply:
Yeah, he did, but that work is incomplete.